hoodie-client/src/main/java/com/uber/hoodie/table/HoodieCopyOnWriteTable.java

```diff
@@ -382,44 +381,40 @@ protected HoodieMergeHandle getUpdateHandle(String commitTime, String fileId,
    */
   @Override
   @SuppressWarnings("unchecked")
-  public Optional<Integer> finalizeWrite(JavaSparkContext jsc, List writeStatuses) {
-    if (!config.shouldUseTempFolderForCopyOnWrite()) {
-      return Optional.empty();
-    }
-
-    // This is to rename each data file from temporary path to its final location
-    List<Tuple2<String, Boolean>> results = jsc
-        .parallelize(writeStatuses, config.getFinalizeWriteParallelism()).map(writeStatus -> {
-          Tuple2<String, HoodieWriteStat> writeStatTuple2 = (Tuple2<String, HoodieWriteStat>)
-              writeStatus;
-          HoodieWriteStat writeStat = writeStatTuple2._2();
-          final FileSystem fs = getMetaClient().getFs();
-          final Path finalPath = new Path(config.getBasePath(), writeStat.getPath());
-
-          if (writeStat.getTempPath() != null) {
-            final Path tempPath = new Path(config.getBasePath(), writeStat.getTempPath());
-            boolean success;
-            try {
-              logger.info("Renaming temporary file: " + tempPath + " to " + finalPath);
-              success = fs.rename(tempPath, finalPath);
-            } catch (IOException e) {
-              throw new HoodieIOException(
-                  "Failed to rename file: " + tempPath + " to " + finalPath);
-            }
+  public void finalizeWrite(JavaSparkContext jsc, List<WriteStatus> writeStatuses)
+      throws HoodieIOException {
+
+    super.finalizeWrite(jsc, writeStatuses);
+
+    if (config.shouldUseTempFolderForCopyOnWrite()) {
+      // This is to rename each data file from temporary path to its final location
+      jsc.parallelize(writeStatuses, config.getFinalizeWriteParallelism())
+          .map(status -> status.getStat())
+          .foreach(writeStat -> {
+            final FileSystem fs = getMetaClient().getFs();
+            final Path finalPath = new Path(config.getBasePath(), writeStat.getPath());
+
+            if (writeStat.getTempPath() != null) {
+              final Path tempPath = new Path(config.getBasePath(), writeStat.getTempPath());
+              boolean success;
+              try {
+                logger.info("Renaming temporary file: " + tempPath + " to " + finalPath);
+                success = fs.rename(tempPath, finalPath);
```

Comment:
Do we need consistency check after rename here ? Isn't rename also eventually consistent ?

Reply:
yes. but I don't think we can recommend people set that since in s3, rename is also a copy -whcih will increase ingest time.. I am planning to fix that once for all in a follow on fix and add any more consistency checks there in.. sound good?